core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -141,6 +148,89 @@ public String getDetails() {
         return this.details;
     }
 
+    /**
+     * The allocation delay value associated with the index (defaulting to node settings if not set).
+     */
+    public long getAllocationDelayTimeoutSetting(Settings settings, Settings indexSettings) {
+        if (reason != Reason.NODE_LEFT) {
+            return 0;
+        }
+        TimeValue delayTimeout = indexSettings.getAsTime(DELAYED_NODE_LEFT_TIMEOUT, settings.getAsTime(DELAYED_NODE_LEFT_TIMEOUT, DEFAULT_DELAYED_NODE_LEFT_TIMEOUT));
+        return Math.max(0l, delayTimeout.millis());
+    }
+
+    /**
+     * The delay in millis when delaying assigning the shard need to expire in.
+     */
+    public long getDelayAllocationExpirationIn(Settings settings, Settings indexSettings) {
+        long delayTimeout = getAllocationDelayTimeoutSetting(settings, indexSettings);
+        if (delayTimeout == 0) {
+            return 0;
+        }
+        long delta = System.currentTimeMillis() - timestamp;
+        // account for time drift, treat it as no timeout
+        if (delta < 0) {
+            return 0;
+        }
+        return delayTimeout - delta;
+    }
+
+
+    /**
+     * Returns the number of shards that are unassigned and currently being delayed.
+     */
+    public static int getNumberOfDelayedUnassigned(Settings settings, ClusterState state) {
+        int count = 0;
+        for (ShardRouting shard : state.routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)) {
+            if (shard.primary() == false) {
+                IndexMetaData indexMetaData = state.metaData().index(shard.getIndex());
+                long delay = shard.unassignedInfo().getDelayAllocationExpirationIn(settings, indexMetaData.getSettings());
+                if (delay > 0) {
+                    count++;
+                }
+            }
+        }
+        return count;
+    }
+
+    /**
+     * Finds the next (closest) delay expiration setting of an unassigned shard. Returns 0 if there are none.
+     */
+    public static long findNextDelayedAllocationSetting(Settings settings, ClusterState state) {
```

Comment:
Minor: can we call this findSmallestDelayedAllocationSettings? Next confused me implying it somehow depends on now.